### PR TITLE
Peeking unknown rooms

### PR DESCRIPTION
**Pull request by ashfame**
PR meant for review (not for merge)

### CODE REVIEW

File: src/domain/session/SessionViewModel.js [hunk 1]
```diff
@@ -218,11 +218,25 @@ export class SessionViewModel extends ViewModel {
         return null;
     }
 
-    _createUnknownRoomViewModel(roomIdOrAlias) {
-        return new UnknownRoomViewModel(this.childOptions({
+    async _createUnknownRoomViewModel(roomIdOrAlias) {
+        let roomId;
+        if ( roomIdOrAlias[0] !== "!" ) {
```
Comment (psrpinto) [inline src/domain/session/SessionViewModel.js]: We could move the room resolution logic to the `canPeekInRoom()` method of `Session`. The `canPeekInRoom()` method would take `roomIdOrAlias` as argument instead of `roomId`, which would be consistent with other methods in `Session`, like `joinRoom(roomIdOrAlias)`.

This would simplify usage of `canPeekInRoom()`, removing the need for the caller to resolve room aliases everywhere where that method would be called.

File: src/platform/web/ui/session/room/timeline/BaseMessageView.js [hunk 2]
```diff
@@ -62,7 +62,7 @@ export class BaseMessageView extends TemplateView {
                 li.removeChild(li.querySelector(".Timeline_messageAvatar"));
                 li.removeChild(li.querySelector(".Timeline_messageSender"));
             } else if (!isContinuation && !this._isReplyPreview) {
-                const avatar = tag.a({href: vm.memberPanelLink, className: "Timeline_messageAvatar"}, [renderStaticAvatar(vm, 30)]);
+                const avatar = vm.options.peekable ? tag.div({className: "Timeline_messageAvatar"}, [renderStaticAvatar(vm, 30)]) : tag.a({href: vm.memberPanelLink, className: "Timeline_messageAvatar"}, [renderStaticAvatar(vm, 30)]);
```
Comment (psrpinto) [inline src/platform/web/ui/session/room/timeline/BaseMessageView.js]: An alternative way to achieve the same result would be to set the [`pointer-events`](https://developer.mozilla.org/en-US/docs/Web/CSS/pointer-events) CSS property to `none`:

```css
.Timeline_messageAvatar {
    // Prevent user panel from opening when clicking on avatars in the timeline.
    pointer-events: none;
}
```

Since we're already customising the Timeline for peekable rooms in `theme.css`, doing this through CSS as well would make sense, in my opinion.

File: src/platform/web/ui/session/SessionView.js [hunk 3]
```diff
@@ -58,6 +59,8 @@ export class SessionView extends TemplateView {
                         return new RoomView(vm.currentRoomViewModel, viewClassForTile);
                     } else if (vm.currentRoomViewModel.kind === "roomBeingCreated") {
                         return new RoomBeingCreatedView(vm.currentRoomViewModel);
+                    } else if (vm.currentRoomViewModel.kind === "peekable") {
```
Comment (psrpinto) [inline src/platform/web/ui/session/SessionView.js]: Just a detail, but should the value be `peekableRoom` instead of `peekable`? Looking at other kinds above (`invite`, `room`), `peekableRoom` could be more explicit.

File: src/platform/web/ui/session/room/PeekableRoomView.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import {TemplateView} from "../../general/TemplateView";
+import {TimelineView} from "./TimelineView";
+import {TimelineLoadingView} from "./TimelineLoadingView";
+import {AvatarView} from "../../AvatarView";
+
+export class PeekableRoomView extends TemplateView {
```
Comment (psrpinto) [inline src/platform/web/ui/session/room/PeekableRoomView.js]: It feels a bit weird to me that we now have both the `UnknownRoomView` and the `PeekableRoomView` but a single `UnknownRoomViewModel`.

To me what we're achieving in this PR is to add functionality to the screen that the user sees when trying to access a room they're not a member of. This new functionality makes it so that that screen shows the timeline of the room, when said room is peekable. Since it's the same screen, I think it would make sense to have all the logic and view-logic in the same place.

In short, I think it could make sense to include the view-logic in this file into `UnknownRoomView`, so that there is a single view and a single view-model for that screen. Sometimes it makes sense to split view-logic into separate components but, in this case, the resulting unified view would still be very simple so I'm not sure separating it brings advantages.

File: src/domain/session/room/UnknownRoomViewModel.js [hunk 5]
```diff
@@ -15,17 +15,25 @@ limitations under the License.
...
 import {ViewModel} from "../../ViewModel";
+import {TimelineViewModel} from "./timeline/TimelineViewModel";
+import {tileClassForEntry as defaultTileClassForEntry} from "./timeline/tiles/index";
+import {getAvatarHttpUrl} from "../../avatar";
 
 export class UnknownRoomViewModel extends ViewModel {
     constructor(options) {
         super(options);
-        const {roomIdOrAlias, session} = options;
+        const {roomIdOrAlias, session, peekable} = options;
```
Comment (psrpinto) [inline src/domain/session/room/UnknownRoomViewModel.js]: This is a generic comment for the whole PR and not about this specific line.

Looking at the spec for [Room History Visibility](https://spec.matrix.org/v1.5/client-server-api/#room-history-visibility), the term defined there is `world_readable`, instead of `peekable`:

> `world_readable` - All events while this is the `m.room.history_visibility` value may be shared by any participating homeserver with anyone, regardless of whether they have ever joined the room.

I think it's generally best to use in the code the same terms that users use, since users could already be familiar with the spec-defined term. Doing this makes it easier to understand the code, by removing one layer of indirection.

So my suggestion would be to use the term `wordReadable` instead of `peekable` in the entirety of the codebase. Some examples:

```js
get kind() {
    return this._worldReadable ? "worldReadableRoom" : "unknown";
}

this._session.loadWorldReadableRoom(this.roomIdOrAlias);
```

File: src/domain/session/room/UnknownRoomViewModel.js [hunk 6]
```diff
@@ -53,6 +61,38 @@ export class UnknownRoomViewModel extends ViewModel {
...
+
+    get timelineViewModel() {
+        return this._timelineVM;
+    }
+
+    avatarUrl(size) {
+        return getAvatarHttpUrl(this._room.avatarUrl, size, this.platform, this._room.mediaRepository);
+    }
+
+    async peek() {
```
Comment (psrpinto) [inline src/domain/session/room/UnknownRoomViewModel.js]: We could rename this to `load()` so that it's consistent with `RoomViewModel`, where the `load()` method renders the timeline.

File: src/platform/web/ui/session/room/PeekableRoomView.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+                        ])
+                    )]),
+                t.mapView(vm => vm.timelineViewModel, timelineViewModel => {
+                    return timelineViewModel ?
+                        new TimelineView(timelineViewModel, this._viewClassForTile) :
+                        new TimelineLoadingView(vm);    // vm is just needed for i18n
+                }),
+                t.div({className: "PeekableRoomComposerView"}, [
+                    t.h3(vm => vm.i18n`Join the room to participate`),
+                    t.button({className: "joinRoomButton", onClick: () => vm.join()}, vm.i18n`Join Room`)
```
Comment (psrpinto) [inline src/platform/web/ui/session/room/PeekableRoomView.js]: I noticed that on small screens the button wraps to a new line.

Comment (psrpinto) [reply src/platform/web/ui/session/room/PeekableRoomView.js]: I also noticed the button is un-styled, I think I remember seeing styled buttons somewhere in hydrogen but can't be sure.

**Review by psrpinto — COMMENTED** [review]
This is looking great, looking forward to having this feature in hydrogen! Well done!

I noticed that when viewing large rooms there is a long delay between the UI first rendering and the timeline appearing, so I think we would probably need to show a loading state before the timeline actually renders:


https://user-images.githubusercontent.com/550401/218133385-b5441e0f-1c3b-4f65-bca2-36a53356227b.mov

File: src/domain/session/SessionViewModel.js (hunk 1 above)
Comment (ashfame) [reply src/domain/session/SessionViewModel.js]: Great suggestion! ✅

File: src/domain/session/room/UnknownRoomViewModel.js (hunk 6 above)
Comment (ashfame) [reply src/domain/session/room/UnknownRoomViewModel.js]: For regular rooms, `load` is a straight forward choice since its only loading data from indexedDB table(s). Here we need to fetch, store and then load, so something like `fetchAndLoad` might be more appropriate, but I have changed it to `load` for now. ✅

File: src/platform/web/ui/session/room/PeekableRoomView.js (hunk 7 above)
Comment (ashfame) [reply src/platform/web/ui/session/room/PeekableRoomView.js]: On the smallest width of screens it only wraps to a new line if you are zoomed in inside of your browser tab. Not sure what would be a fix for that.

I kept the button unstyled as I that's what we have on settings page. All buttons are just default browser styles.

File: src/platform/web/ui/session/room/timeline/BaseMessageView.js (hunk 2 above)
Comment (ashfame) [reply src/platform/web/ui/session/room/timeline/BaseMessageView.js]: Cool! ✅

File: src/platform/web/ui/session/room/PeekableRoomView.js (hunk 4 above)
Comment (ashfame) [reply src/platform/web/ui/session/room/PeekableRoomView.js]: Agree! I moved the code and now everything is handled by `UnknownRoomView` ✅

File: src/domain/session/room/UnknownRoomViewModel.js (hunk 5 above)
Comment (ashfame) [reply src/domain/session/room/UnknownRoomViewModel.js]: Good idea! I updated that throughout the PR ✅

File: src/platform/web/ui/session/SessionView.js (hunk 3 above)
Comment (ashfame) [reply src/platform/web/ui/session/SessionView.js]: Updated it to `worldReadableRoom` ✅

**Comment by ashfame** [comment]
@psrpinto Thanks for the great suggestions! <3

Could you explain the issue in your screencast? `matrix.org` doesn't allow previews on their rooms if you are not in the room and hence peeking can't be functional for those rooms. I believe they must have done it to prevent DOS attacks or to improve scaling. How were you able to even trigger the timeline to load for you?

File: src/domain/session/room/UnknownRoomViewModel.js [hunk 8]
```diff
@@ -53,6 +61,38 @@ export class UnknownRoomViewModel extends ViewModel {
...
+        return getAvatarHttpUrl(this._room.avatarUrl, size, this.platform, this._room.mediaRepository);
+    }
+
+    async load() {
+        if (!this._worldReadable) {
+            return;
+        }
+        try {
+            this._room = await this._session.loadWorldReadableRoom(this.roomIdOrAlias);
+            const timeline = await this._room.openTimeline();
```
Comment (bwindels) [inline src/domain/session/room/UnknownRoomViewModel.js:92]: Interesting ... how does this work if you don't already have a fragment of the timeline stored? If I understand things correctly, at this point, the room exists solely in-memory and nothing related has been written to storage, right? So what token do you end up calling /messages with? I might very well be missing something :)

Comment (ashfame) [reply src/domain/session/room/UnknownRoomViewModel.js:92]: Actually `loadWorldReadableRoom()` above it does all of that - creates the Room object instance, fetch events from messages endpoint & load it all. You will see that here - https://github.com/Automattic/hydrogen-web/pull/3/files#diff-25d3af8c52953975b54837c62ef3195e1054638339aa9b1a9c1d6bc267cdadc8R962

File: src/matrix/Session.js [hunk 9]
```diff
@@ -942,12 +959,118 @@ export class Session {
...
         return this._platform.logger.wrapOrRun(log, "joinRoom", async log => {
             const body = await this._hsApi.joinIdOrAlias(roomIdOrAlias, {log}).response();
             return body.room_id;
         });
     }
+
+    isWorldReadableRoom(roomIdOrAlias, log = null) {
+        return this._platform.logger.wrapOrRun(log, "isWorldReadableRoom", async log => {
+            let roomId;
+            if (roomIdOrAlias[0] !== "!") {
```
Comment (psrpinto) [inline src/matrix/Session.js]:
```suggestion
            if (!roomIdOrAlias.startsWith("!")) {
```

File: src/platform/web/ui/session/room/PeekableRoomView.js (hunk 7 above)
Comment (psrpinto) [reply src/platform/web/ui/session/room/PeekableRoomView.js]: I can no longer reproduce the issue of the button wrapping to a new line, please ignore.

**Review by psrpinto — COMMENTED** [review]
> Could you explain the issue in your screencast?
> How were you able to even trigger the timeline to load for you?

In the screen recording I'm using `#matrix:matrix.org` as an example (`!OGEhHVWSdvArJzumhm:matrix.org`). I'm not a member of that room but I think I was a member and then left, that's why the last message on the timeline is `@psrpinto:matrix.org left the room`. 

So that might explain why the timeline renders for me, which would mean that we would have to take that scenario in account.

But what should we do? Should we:

1. Render the `UnknownRoomView` and show the latest messages
2. Don't render the `UnknownRoomView` and let hydrogen take care of it. Hydrogen already shows the timeline if you left the room, I believe.

At the moment we're in the middle of 1. and 2., where we show the `UnknownRoomView` with not the most recent messages.

> `matrix.org` doesn't allow previews on their rooms if you are not in the room and hence peeking can't be functional for those rooms.

By "doesn't allow" you mean that you can't call the `/messages` endpoint for the room if you're not a member?

**Comment by psrpinto** [comment]
> Hydrogen already shows the timeline if you left the room, I believe.

Actually, on `master`. if I navigate to `!OGEhHVWSdvArJzumhm%3Amatrix.org` (`#matrix:matrix.org`)  it shows "You are currently not in !OGEhHVWSdvArJzumhm:matrix.org. Want to join it?", it doesn't render the timeline.

Am I going crazy? I  thought I'd seen hydrogen rendering the timeline for rooms you left, with history up to the point you left, and a message in the bottom saying "You left the room".

**Comment by ashfame** [comment]
> So that might explain why the timeline renders for me, which would mean that we would have to take that scenario in account.

That would be the timeline you saw for an archive room. Unless we choose to forget the room, UnknownRoomView doesn't come into the picture and by then peeking functionality handles it.

> By "doesn't allow" you mean that you can't call the /messages endpoint for the room if you're not a member?

Correct! If you try to peek into matrix hq room, you will see console errors stating previews are disabled.

I tried peeking into `!OGEhHVWSdvArJzumhm%3Amatrix.org` and it was 403: `User @p1:synapse.dev:8008 not in room !OGEhHVWSdvArJzumhm:matrix.org, and room previews are disabled` which is shown in console as well.

I think you got confused by the archive room timeline view after leaving the room.

**Comment by ashfame** [comment]
@psrpinto and I discussed this more and it turns out this happens when you try to peek into a room that you earlier were a member in but not anymore. Weirdly enough matrix.org has room previews disabled if you are trying to peek from another homeserver but if you are logged in with an account on matrix.org, it lets you preview but with this edge case of /messages endpoint acting up. I will continue to look into this.

Also while trying to figure this out, I was happy to learn that back scrolling is functional as is :)

**Comment by ashfame** [comment]
Created PR on upstream - https://github.com/vector-im/hydrogen-web/pull/1037 so closing this PR
